File: sources/moz-transformer/download_model.py
#!/usr/bin/env python3
"""
download_model.py -- Download and export a pretrained music model for MOZ'Transformer.

Usage:
  python3 download_model.py <model_name> [output_base_dir]

Available models:
  music_transformer   -- Google Magenta Music Transformer (unconditional, classical piano)
  coconet             -- Google Magenta CocoNet (Bach chorales)

Output base dir defaults to ~/Documents/MOZLib-models/
Each model exports to: <output_base_dir>/<model_name>_weights/

This script is designed to be called from Max's shell object:
  shell "python3 /path/to/download_model.py music_transformer"
  -> prints the weights directory path on the last line for Max to read back

Requires:
  pip install tensorflow    (for Magenta TF1 checkpoints)
  pip install requests      (for downloading)
"""

import sys
import os
import json

MODELS = {
    "music_transformer": {
        "description": "Google Magenta Music Transformer (unconditional, classical piano)",
        "files": [
            {
                "url": "https://storage.googleapis.com/magentadata/models/music_transformer/checkpoints/unconditional_model_16.ckpt.data-00000-of-00001",
                "filename": "unconditional_model_16.ckpt.data-00000-of-00001"
            },
            {
                "url": "https://storage.googleapis.com/magentadata/models/music_transformer/checkpoints/unconditional_model_16.ckpt.index",
                "filename": "unconditional_model_16.ckpt.index"
            },
            {
                "url": "https://storage.googleapis.com/magentadata/models/music_transformer/checkpoints/unconditional_model_16.ckpt.meta",
                "filename": "unconditional_model_16.ckpt.meta"
            }
        ],
        "ckpt_base": "unconditional_model_16.ckpt",
        "format": "tf1"
    },
    "coconet": {
        "description": "Google Magenta CocoNet (Bach chorale harmonization)",
        "files": [
            {
                "url": "https://storage.googleapis.com/magentadata/models/coconet/checkpoint",
                "filename": "checkpoint"
            },
            {
                "url": "https://storage.googleapis.com/magentadata/models/coconet/model.ckpt-0.data-00000-of-00001",
                "filename": "model.ckpt-0.data-00000-of-00001"
            },
            {
                "url": "https://storage.googleapis.com/magentadata/models/coconet/model.ckpt-0.index",
                "filename": "model.ckpt-0.index"
            },
            {
                "url": "https://storage.googleapis.com/magentadata/models/coconet/model.ckpt-0.meta",
                "filename": "model.ckpt-0.meta"
            }
        ],
        "ckpt_base": "checkpoint",
        "format": "tf1"
    }
}


def download_file(url, dest_path):
    """Download a file with progress display."""
    try:
        import requests
    except ImportError:
        print("ERROR: requests not installed. Run: pip install requests")
        sys.exit(1)

    filename = os.path.basename(dest_path)
    print("  Downloading {}...".format(filename))

    try:
        response = requests.get(url, stream=True)
        response.raise_for_status()

        total = int(response.headers.get("content-length", 0))
        downloaded = 0

        with open(dest_path, "wb") as f:
            for chunk in response.iter_content(chunk_size=65536):
                if chunk:
                    f.write(chunk)
                    downloaded += len(chunk)
                    if total > 0:
                        pct = int(downloaded * 100 / total)
                        mb = downloaded / (1024 * 1024)
                        print("\r    {:.1f} MB  {}%".format(mb, pct),
                              end="", flush=True)
        print()  # newline after progress
        print("    -> {}".format(dest_path))

    except Exception as e:
        print("ERROR downloading {}: {}".format(url, e))
        sys.exit(1)


def download_and_export(model_name, output_base_dir=None):
    if model_name not in MODELS:
        print("ERROR: Unknown model '{}'".format(model_name))
        print("Available: {}".format(", ".join(MODELS.keys())))
        sys.exit(1)

    model_info = MODELS[model_name]

    # Default output base dir
    if output_base_dir is None:
        home = os.path.expanduser("~")
        output_base_dir = os.path.join(home, "Documents", "MOZLib-models")

    output_base_dir = os.path.abspath(output_base_dir)
    ckpt_dir = os.path.join(output_base_dir, model_name + "_ckpt")
    weights_dir = os.path.join(output_base_dir, model_name + "_weights")

    os.makedirs(ckpt_dir, exist_ok=True)
    os.makedirs(weights_dir, exist_ok=True)

    print("=" * 60)
    print("MOZ'Transformer model download")
    print("Model:  {}".format(model_name))
    print("Desc:   {}".format(model_info["description"]))
    print("=" * 60)

    # Download checkpoint files
    all_present = True
    for file_info in model_info["files"]:
        dest = os.path.join(ckpt_dir, file_info["filename"])
        if os.path.exists(dest):
            print("  Already downloaded: {}".format(file_info["filename"]))
        else:
            all_present = False
            download_file(file_info["url"], dest)

    if all_present:
        print("All checkpoint files already present, skipping download.")

    # Check if weights already exported
    manifest_path = os.path.join(weights_dir, "manifest.json")
    if os.path.exists(manifest_path):
        print("\nWeights already exported at: {}".format(weights_dir))
        print("\nMOZ_WEIGHTS_DIR:{}".format(weights_dir))
        return weights_dir

    # Export weights
    print("\nExporting weights...")
    ckpt_path = os.path.join(ckpt_dir, model_info["ckpt_base"])

    # Import and run export_weights
    script_dir = os.path.dirname(os.path.abspath(__file__))
    sys.path.insert(0, script_dir)

    try:
        import export_weights as ew
        ew.export_weights(ckpt_path, weights_dir)
    except ImportError:
        print("ERROR: export_weights.py not found in {}".format(script_dir))
        print("Make sure export_weights.py is in the same directory as this script.")
        sys.exit(1)

    print("\nModel ready.")
    print("Weights directory: {}".format(weights_dir))

    # Print machine-readable output for Max to parse
    # Max shell object reads stdout — last line with this prefix gives the path
    print("\nMOZ_WEIGHTS_DIR:{}".format(weights_dir))

    return weights_dir


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(__doc__)
        print("Available models:")
        for name, info in MODELS.items():
            print("  {:20s} {}".format(name, info["description"]))
        sys.exit(1)

    model_name = sys.argv[1]
    output_base_dir = sys.argv[2] if len(sys.argv) > 2 else None
    download_and_export(model_name, output_base_dir)

File: sources/moz-transformer/download-model.js
/**
 * download-model.js — MOZ'Transformer model downloader for Node4Max
 *
 * Télécharge les checkpoints Magenta directement depuis Max.
 *
 * Usage depuis Max:
 *   [node.script download-model.js]
 *   -> envoyer "download music_transformer" ou "download coconet"
 *   -> envoyer "download music_transformer /custom/output/path"
 *
 * Outlets:
 *   0: messages de progression et statut (bang quand terminé)
 *   1: path du checkpoint (à passer à load-model-from-ckpt en Lisp)
 *   2: erreurs
 *
 * Inlet:
 *   "download <model_name> [output_dir]"
 *   "list"  -> liste les modèles disponibles
 */

const Max = require('max-api');
const https = require('https');
const fs = require('fs');
const path = require('path');
const os = require('os');

// -------------------------------------------------------------------------
// Modèles disponibles
// -------------------------------------------------------------------------

const MODELS = {
  music_transformer: {
    description: 'Google Magenta Music Transformer (piano classique, ~85MB)',
    files: [
      'https://storage.googleapis.com/magentadata/models/music_transformer/checkpoints/unconditional_model_16.ckpt.data-00000-of-00001',
      'https://storage.googleapis.com/magentadata/models/music_transformer/checkpoints/unconditional_model_16.ckpt.index',
      'https://storage.googleapis.com/magentadata/models/music_transformer/checkpoints/unconditional_model_16.ckpt.meta'
    ],
    ckpt_base: 'unconditional_model_16.ckpt'
  },
  coconet: {
    description: 'Google Magenta CocoNet (chorals de Bach, ~120MB)',
    files: [
      'https://storage.googleapis.com/magentadata/models/coconet/checkpoint',
      'https://storage.googleapis.com/magentadata/models/coconet/model.ckpt-0.data-00000-of-00001',
      'https://storage.googleapis.com/magentadata/models/coconet/model.ckpt-0.index',
      'https://storage.googleapis.com/magentadata/models/coconet/model.ckpt-0.meta'
    ],
    ckpt_base: 'model.ckpt-0'
  }
};

// -------------------------------------------------------------------------
// Téléchargement d'un fichier avec progression
// -------------------------------------------------------------------------

function downloadFile(url, destPath) {
  return new Promise((resolve, reject) => {
    const filename = path.basename(url);

    // Déjà téléchargé ?
    if (fs.existsSync(destPath)) {
      Max.post(`  déjà présent: ${filename}`);
      resolve();
      return;
    }

    Max.post(`  téléchargement: ${filename}`);

    const tmpPath = destPath + '.tmp';
    const file = fs.createWriteStream(tmpPath);

    const request = https.get(url, (response) => {
      // Gérer les redirections
      if (response.statusCode === 301 || response.statusCode === 302) {
        file.close();
        fs.unlinkSync(tmpPath);
        downloadFile(response.headers.location, destPath)
          .then(resolve)
          .catch(reject);
        return;
      }

      if (response.statusCode !== 200) {
        file.close();
        fs.unlinkSync(tmpPath);
        reject(new Error(`HTTP ${response.statusCode} pour ${url}`));
        return;
      }

      const totalBytes = parseInt(response.headers['content-length'] || '0', 10);
      let downloadedBytes = 0;
      let lastPct = -1;

      response.on('data', (chunk) => {
        downloadedBytes += chunk.length;
        if (totalBytes > 0) {
          const pct = Math.floor(downloadedBytes * 100 / totalBytes);
          // Envoyer progression toutes les 10%
          if (pct !== lastPct && pct % 10 === 0) {
            lastPct = pct;
            const mb = (downloadedBytes / (1024 * 1024)).toFixed(1);
            Max.outlet(0, `progress ${filename} ${pct}% ${mb}MB`);
          }
        }
      });

      response.pipe(file);

      file.on('finish', () => {
        file.close(() => {
          fs.renameSync(tmpPath, destPath);
          Max.post(`  -> ${destPath}`);
          resolve();
        });
      });
    });

    request.on('error', (err) => {
      file.close();
      if (fs.existsSync(tmpPath)) fs.unlinkSync(tmpPath);
      reject(err);
    });

    file.on('error', (err) => {
      file.close();
      if (fs.existsSync(tmpPath)) fs.unlinkSync(tmpPath);
      reject(err);
    });
  });
}

// -------------------------------------------------------------------------
// Téléchargement d'un modèle complet
// -------------------------------------------------------------------------

async function downloadModel(modelName, outputDir) {
  const info = MODELS[modelName];
  if (!info) {
    const available = Object.keys(MODELS).join(', ');
    Max.outlet(2, `modèle inconnu: ${modelName} (disponibles: ${available})`);
    return;
  }

  // Répertoire de sortie par défaut
  if (!outputDir) {
    outputDir = path.join(os.homedir(), 'ml', 'MOZLib-models', modelName);
  }

  // Créer le répertoire si nécessaire — sans jamais effacer ce qui existe
  const mlDir       = path.join(os.homedir(), 'ml');
  const modelsDir   = path.join(mlDir, 'MOZLib-models');

  for (const dir of [mlDir, modelsDir, outputDir]) {
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir);
      Max.post(`  créé: ${dir}`);
    } else {
      Max.post(`  existe déjà: ${dir}`);
    }
  }

  Max.outlet(0, `début du téléchargement: ${modelName}`);
  Max.post(`=== MOZ'Transformer download ===`);
  Max.post(`Modèle: ${modelName}`);
  Max.post(`Desc:   ${info.description}`);
  Max.post(`Dest:   ${outputDir}`);

  try {
    // Télécharger chaque fichier séquentiellement
    for (let i = 0; i < info.files.length; i++) {
      const url = info.files[i];
      const filename = path.basename(url);
      const destPath = path.join(outputDir, filename);

      Max.outlet(0, `fichier ${i + 1}/${info.files.length}: ${filename}`);
      await downloadFile(url, destPath);
    }

    const ckptPath = path.join(outputDir, info.ckpt_base);

    Max.post(`\nTéléchargement terminé.`);
    Max.post(`Charger avec:`);
    Max.post(`  (moz-transformer:load-model-from-ckpt "${ckptPath}")`);

    // Outlet 0: signal de fin
    Max.outlet(0, 'done');
    // Outlet 1: le path du checkpoint pour Max/Lisp
    Max.outlet(1, ckptPath);

  } catch (err) {
    Max.post(`ERREUR: ${err.message}`);
    Max.outlet(2, `erreur: ${err.message}`);
  }
}

// -------------------------------------------------------------------------
// Handlers Max
// -------------------------------------------------------------------------

Max.addHandler('download', (...args) => {
  const modelName = args[0];
  const outputDir = args[1] || null;

  if (!modelName) {
    Max.outlet(2, 'usage: download <model_name> [output_dir]');
    return;
  }

  downloadModel(modelName, outputDir);
});

Max.addHandler('list', () => {
  Max.post('=== Modèles disponibles ===');
  for (const [name, info] of Object.entries(MODELS)) {
    Max.post(`  ${name}: ${info.description}`);
    Max.outlet(0, `model ${name}`);
  }
});

// -------------------------------------------------------------------------
// Démarrage
// -------------------------------------------------------------------------

Max.post('MOZ\'Transformer download script prêt.');
Max.post('Envoyer "download music_transformer" ou "download coconet"');
Max.outlet(0, 'ready');
